File: src/main/java/features/numinfofromimage/ImageSeriesVoxelVolumeCalcAlg.java
package features.numinfofromimage;

import core.algorithm.Algorithm;
import core.data.ParametrizedData;
import core.data.attribute.MirfAttributes;
import core.data.medimage.ImageSeriesData;
import features.dicomimage.data.DicomAttributes;

/**
 * Calculates the volume of the given {@link ImageSeriesData}.
 */
public class ImageSeriesVoxelVolumeCalcAlg implements Algorithm<ImageSeriesData, ParametrizedData> {

    @Override
    public ParametrizedData execute(ImageSeriesData input) {

        double result = 0;

        for (var image : input.getImages()) {
            //TODO: (avlomakin) replace with Log.Warning
            if (!image.getAttributes().hasAttribute(MirfAttributes.INSTANCE.getTHRESHOLDED()))
                System.out.println("Warning, performing volume calculation on non-thresholded image, possible unexpected result");

            double onePixelVolume = image.getAttributes().findAttributeValue(DicomAttributes.ONE_PIXEL_VOLUME);

            var img = image.getImage();
            var raw = img.getRaster().getPixels(img.getMinX(), img.getMinY(), img.getWidth(), img.getHeight(), (int[]) null);

            for (var aLine : raw) result += (aLine != 0) ? onePixelVolume : 0;
        }
        return new ParametrizedData<>(result);
    }
}

File: src/main/java/features/numinfofromimage/ImageSeriesVoxelVolumeCalcAlg.kt
package features.numinfofromimage

import core.algorithm.Algorithm
import core.data.ParametrizedData
import core.data.attribute.MirfAttributes
import core.data.medimage.ImageSeriesData
import features.dicomimage.data.DicomAttributes

/**
 * Calculates the volume of the given [ImageSeriesData].
 */
class ImageSeriesVoxelVolumeCalcAlg : Algorithm<ImageSeriesData, ParametrizedData<*>> {

    override fun execute(input: ImageSeriesData): ParametrizedData<*> {

        var result = 0.0

        for (image in input.images) {
            //TODO: (avlomakin) replace with Log.Warning
            if (!image.attributes.hasAttribute(MirfAttributes.THRESHOLDED))
                println("Warning, performing volume calculation on non-thresholded image, possible unexpected result")

            val onePixelVolume = image.attributes.findAttributeValue(DicomAttributes.ONE_PIXEL_VOLUME)!!

            val img = image.image
            val raw = img!!.raster.getPixels(img.minX, img.minY, img.width, img.height, null as IntArray?)

            for (aLine in raw)
                result += if (aLine != 0) onePixelVolume else 0.0

        }
        return ParametrizedData(result)
    }
}
